### PR TITLE
[analytics] improve serialization support for maps

### PR DESCRIPTION
**Pull request by dcramer**
This allows you to pass instances as attribute values.

```
>>> analytics.record('subscription.changed', current=Subscription())
```

#nochanges

### CODE REVIEW

File: src/sentry/analytics/event.py [hunk 1]
```diff
@@ -9,23 +9,43 @@
...
         self.required = required
-        self.attributes = attributes
+        if attributes:
+            self.attributes = attributes
+
+    def serialize(self, value):
+        """
+        If passed a non dictionary we assume we can pull attributes from it.
+
+        This will hard error in some situations if you're passing some obscure
```
Comment (tkaemming) [inline src/sentry/analytics/event.py]: An integer might be unexpected here, but it's not obscure

File: src/sentry/analytics/event.py [hunk 2]
```diff
@@ -9,23 +9,43 @@
...
 
 class Attribute(object):
-    __slots__ = ['name', 'type', 'required']
-
     def __init__(self, name, type=six.text_type, required=True):
         self.name = name
         self.type = type
         self.required = required
 
+    def serialize(self, value):
```
Comment (tkaemming) [inline src/sentry/analytics/event.py]: This isn't serialization

File: src/sentry/analytics/event.py [hunk 3]
```diff
@@ -9,23 +9,43 @@
...
 
 class Map(Attribute):
-    def __init__(self, name, attributes, required=True):
+    attributes = ()
+
+    def __init__(self, name, attributes=None, required=True):
         self.name = name
         self.required = required
-        self.attributes = attributes
+        if attributes:
```
Comment (tkaemming) [inline src/sentry/analytics/event.py]: I don't understand the motivation behind this change and the one on L24

File: src/sentry/analytics/event.py [hunk 4]
```diff
@@ -9,23 +9,43 @@
...
+        This will hard error in some situations if you're passing some obscure
+        value (like an int).
+        """
+        if value is None:
+            return value
 
-    def type(self, value):
         if not isinstance(value, Mapping):
-            raise ValueError('Value must be a dictionary')
+            new_value = {}
```
Comment (tkaemming) [inline src/sentry/analytics/event.py]: Seems like it'd make sense to make L53 an `else` and just assign this to `items` outright

**Review by tkaemming — APPROVED** [review]

File: src/sentry/analytics/event.py (hunk 3 above)
Comment (dcramer) [inline src/sentry/analytics/event.py]: might not need this, but i was trying to move in the direction of having Event inherit from Map so I could just (more easily) reuse what is more or less duplicated code

File: src/sentry/analytics/event.py (hunk 2 above)
Comment (dcramer) [inline src/sentry/analytics/event.py]: what would you call it?

(i would argue it is serialization, as its serializing to a specified format)

Comment (dcramer) [inline src/sentry/analytics/event.py]: actually i could just leave it as type, but that was kind of a hack

Comment (tkaemming) [inline src/sentry/analytics/event.py]: I'm not sure of a name offhand ("extract" might be a better verb but this has more responsibilities than just extracting values from the input that at this point, incl. type validation).

The "serialization" part here is really just converting the input type to a different output type and *preparing* it for serialization, but serialization is [more or less the act of converting a data structure into a byte stream](https://en.wikipedia.org/wiki/Serialization) that can be transmitted/stored serially. (Our API "serializer" isn't serialization either.)

**Pull request merged**